pkg/reconciler/apis/apiexport/apiexport_controller.go

```diff
@@ -0,0 +1,383 @@
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return nil // object deleted before we handled it
+		}
+		return err
+	}
+
+	old := obj
+	obj = obj.DeepCopy()
+
+	reconcileErr := c.reconcile(ctx, obj)
+
+	// Regardless of whether reconcile returned an error or not, always try to patch status if needed. Return the
+	// reconciliation error at the end.
+
+	// If the object being reconciled changed as a result, update it.
+	if !equality.Semantic.DeepEqual(old.Status, obj.Status) {
+		oldData, err := json.Marshal(apisv1alpha1.APIExport{
+			Status: old.Status,
+		})
+		if err != nil {
+			return fmt.Errorf("failed to Marshal old data for apibinding %s|%s: %w", clusterName, name, err)
+		}
+
+		newData, err := json.Marshal(apisv1alpha1.APIExport{
+			ObjectMeta: metav1.ObjectMeta{
+				UID:             old.UID,
+				ResourceVersion: old.ResourceVersion,
+			}, // to ensure they appear in the patch as preconditions
+			Status: obj.Status,
+		})
+		if err != nil {
+			return fmt.Errorf("failed to Marshal new data for apibinding %s|%s: %w", clusterName, name, err)
+		}
+
+		patchBytes, err := jsonpatch.CreateMergePatch(oldData, newData)
+		if err != nil {
+			return fmt.Errorf("failed to create patch for apiexport %s|%s: %w", clusterName, name, err)
+		}
+		_, uerr := c.kcpClusterClient.Cluster(clusterName).ApisV1alpha1().APIExports().Patch(ctx, obj.Name, types.MergePatchType, patchBytes, metav1.PatchOptions{}, "status")
+		return uerr
+	}
+
+	return reconcileErr
+}
+
+func (c *controller) reconcile(ctx context.Context, apiExport *apisv1alpha1.APIExport) error {
+	identity := apiExport.Spec.Identity
+	if identity == nil {
+		identity = &apisv1alpha1.Identity{}
+	}
+
+	clusterName := logicalcluster.From(apiExport)
+
+	if identity.SecretRef == nil {
+		if _, err := c.namespaceLister.Get(clusters.ToClusterAwareKey(clusterName, c.secretNamespace)); errors.IsNotFound(err) {
+			ns := &corev1.Namespace{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: c.secretNamespace,
+				},
+			}
+			_, err := c.kubeClusterClient.Cluster(clusterName).CoreV1().Namespaces().Create(ctx, ns, metav1.CreateOptions{})
+			if err != nil && !errors.IsAlreadyExists(err) {
+				klog.Errorf("Error creating namespace %q in cluster %q for APIExport secret identities: %v", c.secretNamespace, clusterName, err)
+				// Keep going - maybe things will work. If the secret creation fails, we'll make sure to set a condition.
+			}
+		}
+
+		privateKey, err := rsa.GenerateKey(cryptorand.Reader, 4096)
+		if err != nil {
+			conditions.MarkFalse(
+				apiExport,
+				apisv1alpha1.APIExportIdentityValid,
+				apisv1alpha1.IdentityGenerationFailedReason,
+				conditionsv1alpha1.ConditionSeverityError,
+				"Error generating identity private key: %v",
+				err,
+			)
+
+			return fmt.Errorf("error generating APIExport %s|%s identity private key: %w", clusterName, apiExport.Name, err)
+		}
+
+		encoded, err := keyutil.MarshalPrivateKeyToPEM(privateKey)
```

Comment:
technically we only need a random secret. An RSA keht would allow us to sign, if we want to. Still wondering how (other than cross-shard SubjectAccessReview) we make a shard to accept wildcard reads. Maybe signing could help (also note that a service account has the workspace today, and with that at least can be checked offline by a shard).

Reply:
Make me think whether we need some generic authorization proof concept in SA JWTs. The bound service tokens actually could give us that. The APIExport hash could be an audience in the JWT.